pipeline_dp/contribution_bounders.py

```diff
@@ -0,0 +1,145 @@
+              accumulator).
+
+        """
+
+
+class SamplingCrossAndPerPartitionContributionBounder(ContributionBounder):
+    """Bounds the contribution by privacy_id per and cross partitions.
+
+    It ensures by performing sampling if needed that each privacy_id contributes
+    not more than to max_partitions_contributed partitions (cross-partition
+    contribution bounding) and not more than max_contributions_per_partition to
+    each contributed partition (per partition contribution bounding).
+    """
+
+    def bound_contributions(self, col, params, backend, report_generator,
+                            aggregate_fn):
+        """See docstrings for this class and for the base class bound_contributions."""
+        max_partitions_contributed = params.max_partitions_contributed
+        max_contributions_per_partition = params.max_contributions_per_partition
+        col = backend.map_tuple(
+            col, lambda pid, pk, v: ((pid, pk), v),
+            "Rekey to ( (privacy_id, partition_key), value))")
+        col = backend.sample_fixed_per_key(
+            col, params.max_contributions_per_partition,
+            "Sample per (privacy_id, partition_key)")
+        report_generator.add_stage(
+            f"Per-partition contribution bounding: randomly selected not "
+            f"more than {max_contributions_per_partition} contributions")
+        # ((privacy_id, partition_key), [value])
+        col = backend.map_values(
+            col, aggregate_fn,
+            "Apply aggregate_fn after per partition bounding")
+        # ((privacy_id, partition_key), accumulator)
+        # Cross partition bounding
+        col = backend.map_tuple(
+            col, lambda pid_pk, v: (pid_pk[0], (pid_pk[1], v)),
+            "Rekey to (privacy_id, (partition_key, accumulator))")
+        col = backend.sample_fixed_per_key(col,
+                                           params.max_partitions_contributed,
+                                           "Sample per privacy_id")
+
+        report_generator.add_stage(
+            f"Cross-partition contribution bounding: randomly selected not more"
+            f" than {max_partitions_contributed} partitions per privacy id")
+
+        # (privacy_id, [(partition_key, accumulator)])
+        def unnest_cross_partition_bound_sampled_per_key(pid_pk_v):
+            pid, pk_values = pid_pk_v
+            return (((pid, pk), v) for (pk, v) in pk_values)
+
+        return backend.flat_map(col,
+                                unnest_cross_partition_bound_sampled_per_key,
+                                "Unnest")
+
+
+class SamplingPerPrivacyIdContributionBounder(ContributionBounder):
+    """Bounds the total contributions of a privacy_id.
+
+    If a privacy_id contributes more than max_contributions, then
+    max_contributions contributions are uniformly sampled, otherwise all
+    contributions are kept.
+
+    Warning: This contribution bounder does group_by by privacy_id and processes
```

Comment:
I am not sure if that'll be a problem actually. If runtime of backend.sample_fixed_per_key does not scale by number of elements per key (and it would be possible to implement it as such), then after that step we only have max_contributions amount of data per privacy id, which should be fairly low.

I thinkg processing of all data per privacy_id is required if we want to do a scaling-based sampling (i.e. found total number of contributions, scale each contribution by that amount).